libmamba/src/core/transaction.cpp

```diff
@@ -825,208 +828,227 @@ namespace mamba
+                );
+                extract_trackers.push_back(task->get_future());
+                download_requests.push_back(fit->build_download_request(
+                    [extract_task = std::move(task)](std::size_t downloaded_size)
                     {
-                        auto active_tasks = dl_bar->active_tasks().size();
-                        if (active_tasks == 0)
-                        {
-                            repr.prefix.set_value(fmt::format("{:<16}", "Downloading"));
-                            repr.postfix.set_value(fmt::format("{:<25}", ""));
-                        }
-                        else
-                        {
-                            repr.prefix.set_value(fmt::format(
-                                "{:<11} {:>4}",
-                                "Downloading",
-                                fmt::format("({})", active_tasks)
-                            ));
-                            repr.postfix.set_value(fmt::format("{:<25}", dl_bar->last_active_task()));
-                        }
-                        repr.current.set_value(fmt::format(
-                            "{:>7}",
-                            to_human_readable_filesize(double(dl_bar->current()), 1)
-                        ));
-                        repr.separator.set_value("/");
-
-                        std::string total_str;
-                        if (dl_bar->total() == std::numeric_limits<std::size_t>::max())
-                        {
-                            total_str = "??.?MB";
-                        }
-                        else
-                        {
-                            total_str = to_human_readable_filesize(double(dl_bar->total()), 1);
-                        }
-                        repr.total.set_value(fmt::format("{:>7}", total_str));
-
-                        auto speed = dl_bar->avg_speed(std::chrono::milliseconds(500));
-                        repr.speed.set_value(
-                            speed
-                                ? fmt::format("@ {:>7}/s", to_human_readable_filesize(double(speed), 1))
-                                : ""
+                        MainExecutor::instance().schedule(
+                            [t = std::move(extract_task)](std::size_t ds) { (*t)(ds); },
+                            downloaded_size
                         );
                     }
-                );
+                ));
             }
+            return download_requests;
+        }
 
-            auto* extract_bar = aggregated_pbar_manager.aggregated_bar("Extract");
-            if (extract_bar)
+        void schedule_extractions(
+            ExtractTaskList& extract_tasks,
+            ExtractTrackerList& extract_trackers,
+            std::size_t download_size
+        )
+        {
+            for (auto it = extract_tasks.begin() + static_cast<std::ptrdiff_t>(download_size);
```

Comment:
Prefer using `std::next`? Also, you're not guaranteed that `.begin()` gives a pointer (see debug build in msvc as an example).

Actually it's not clear to me why `download_size` is used here? Is it actually an index?

Reply:
> Also, you're not guaranteed that .begin() gives a pointer

For random access iterators, we are guaranteed the `it + offset` is valid.

> Actually it's not clear to me why download_size is used here? Is it actually an index?

Wd want to scheldule extractions for packages that don't need to be downloaded, because downloading a package already triggers its extraction.

Reply:
> For random access iterators, we are guaranteed the it + offset is valid.

Ok, then my question seems to be more like why the cast to `ptrdiff_t`? `size_t` doesnt work here?

> We want to scheldule extractions for packages that don't need to be downloaded, because downloading a package already triggers its extraction.

I think it's just the name `download_size` with a `size_t ` as type that's misleading to me?

Reply:
`operator+` for random access iterators accept a `ptrdiff_t` only, using a `size_t` results in a conversion warning.

Reply:
Ok got it. 👍🏽